File: src/actions/print.js
"use strict";

const chalk = require("chalk");
const fe = require("file-exists");
const { exit, paths, read, fmt } = require("../utils/toolbox.js");

module.exports = async (args) => {
	let { registrypath } = paths;
	let { command } = args;
	// eslint-disable-next-line no-unused-vars
	let err, res;
	let tstring = "";

	// Source must be provided.
	if (!command) {
		tstring = "Please provide a command name using the ? flag.";
		exit([fmt(tstring, chalk.bold("--command"))]);
	}

	// If command is supplied then print its acdef/config file contents.
	if (command) {
		// Break apart command.
		let [, cmdname = ""] = command.match(/^(.*?)(\.(acdef))?$/);
		let ext = ".acdef";

		// Exit and give error if a command name not provided.
		if (!cmdname) {
			tstring = "Please provide a command name (i.e. --command=?).";
			exit([fmt(tstring, chalk.bold("nodecliac.acdef"))]);
		}

		// Check if command chain contains invalid characters.
		let r = /[^-._:a-zA-Z0-9\\/]/;
		if (r.test(cmdname)) {
			// Loop over command chain to highlight invalid character.
			let chars = [];
			let invalid_char_count = 0;
			for (let i = 0, l = cmdname.length; i < l; i++) {
				let char = cmdname[i];

				// If an invalid character highlight.
				if (r.test(char)) {
					chars.push(chalk.bold.red(char));
					invalid_char_count++;
				} else chars.push(char);
			}

			// Plural output character string.
			let char_string = `character${invalid_char_count > 1 ? "s" : ""}`;

			// Invalid escaped command-flag found.
			let varg1 = chalk.bold("Invalid:");
			let varg3 = chars.join("");
			exit([fmt("? ? in command: ?", varg1, char_string, varg3)]);
		}

		// File paths.
		let pathstart = `${registrypath}/${cmdname}`;
		let filepath = `${pathstart}/${cmdname}${ext}`;
		let filepathconfig = `${pathstart}/.${cmdname}.config${ext}`;

		// Check if acdef file exists. Print file contents.
		if (await fe(filepath)) {
			// Log file contents.
			console.log(`\n[${chalk.bold(`${cmdname}${ext}`)}]\n`);
			console.log(await read(filepath));

			// Check if config file exists. Print file contents.
			if (await fe(filepathconfig)) {
				// Log file contents.
				let header = chalk.bold(`.${cmdname}.config${ext}`);
				console.log(`[${header}]\n`);
				console.log(await read(filepathconfig));
			}
		} else {
			// If acdef file does not exist log a message and exit script.
			let bcmdname = chalk.bold(cmdname);
			exit([`acdef file for command ${bcmdname} does not exist.`]);
		}
	}
};

File: src/parser/nim/actions/print.nim
proc print() {.async.} =
    var tstring = ""

    # Source must be provided.
    if prcommand.len == 0:
        tstring = "Please provide a command name using the $1 flag."
        quit(tstring % ["--command".chalk("bold")])

    # If command is supplied then print its acdef/config file contents.
    if prcommand.len != 0:
        # Break apart command.
        var cmdname = ""
        if prcommand =~ re"^(.*?)(\.(acdef))?$": cmdname = matches[0]
        let ext = ".acdef"

        # Exit and give error if a command name not provided.
        if cmdname.len == 0:
            tstring = "Please provide a command name (i.e. --command=$1)."
            quit(tstring % ["nodecliac.acdef".chalk("bold")])

        # Check if command chain contains invalid characters.
        let INVALID_CHARS = {'^', '-', '.', '_', ':', '\\', '/'} + Letters + Digits
        if cmdname.find(INVALID_CHARS) != -1:
            # Loop over command chain to highlight invalid character.
            var chars: seq[string] = @[]
            var invalid_char_count = 0
            for i, `char` in cmdname:
                let `char` = cmdname[i]

                # If an invalid character highlight.
                if `char` in INVALID_CHARS:
                    chars.add(($`char`).chalk("bold", "red"))
                    inc(invalid_char_count)
                else: chars.add($`char`)

            # Plural output character string.
            let char_string = "character" & (if invalid_char_count > 1: "s" else: "")

            # Invalid escaped command-flag found.
            let varg1 = "Invalid:".chalk("bold")
            let varg3 = chars.join("")
            quit("$1 $2 in command: $3" % [varg1, char_string, varg3])

        # File paths.
        let pathstart = fmt"{registrypath}/{cmdname}"
        let filepath = fmt"{pathstart}/{cmdname}{ext}"
        let filepathconfig = fmt"{pathstart}/.{cmdname}.config{ext}"

        # Check if acdef file exists. Print file contents.
        if fileExists(filepath):
            # Log file contents.
            echo "\n" & fmt"{cmdname}{ext}".chalk("bold") & "\n"
            echo readFile(filepath)

            # Check if config file exists. Print file contents.
            if fileExists(filepathconfig):
                # Log file contents.
                let header = fmt".{cmdname}.config{ext}".chalk("bold")
                echo fmt"[{header}]\n"
                echo readFile(filepathconfig)
        else:
            # If acdef file does not exist log a message and exit script.
            let bcmdname = cmdname.chalk("bold")
            quit(fmt"acdef file for command {bcmdname} does not exist.")
